ciao-controller/internal/datastore/datastore.go

```diff
@@ -681,7 +681,18 @@ func (ds *Datastore) ReleaseTenantIP(tenantID string, ip string) error {
 	ds.tenantsLock.Lock()
 
 	if ds.tenants[tenantID] != nil {
-		ds.tenants[tenantID].network[int(subnetInt)][int(ipBytes[3])] = false
+		delete(ds.tenants[tenantID].network[int(subnetInt)], int(ipBytes[3]))
+		subnets := ds.tenants[tenantID].subnets
+		network := ds.tenants[tenantID].network
+		i := int(subnetInt)
```

Comment:
Why not use your intermediate variables here:

delete(network[i], int(ipBytes[3]))